src/mon/OSDMonitor.cc

```diff
@@ -10158,6 +10158,49 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
     pending_inc.new_primary_temp[pgid] = osd;
     ss << "set " << pgid << " primary_temp mapping to " << osd;
     goto update;
+  } else if (prefix == "pg repeer") {
+    pg_t pgid;
+    string pgidstr;
+    cmd_getval(cct, cmdmap, "pgid", pgidstr);
+    if (!pgid.parse(pgidstr.c_str())) {
+      ss << "invalid pgid '" << pgidstr << "'";
+      err = -EINVAL;
+      goto reply;
+    }
+    if (!osdmap.pg_exists(pgid)) {
+      ss << "pg '" << pgidstr << "' does not exist";
+      err = -ENOENT;
+      goto reply;
+    }
+    vector<int> acting;
+    int primary;
+    osdmap.pg_to_acting_osds(pgid, &acting, &primary);
+    if (osdmap.pg_temp->count(pgid)) {
+      // remove pg_temp
+      pending_inc.new_pg_temp[pgid].clear();
```

Comment:
i think we cannot just nuke the `new_pg_temp[pgid]` in hope to re-trigger the peering, see https://github.com/ceph/ceph/commit/4c7df944c7f28232873ba681eedce72cdb062ea5#diff-fa6c2eba8356ae1442d1bf749beacfdfR4190 .

Reply:
I saw that, but I couldn't figure out why that would be true.  I updated the logic a bit, though, so that both/all cases involve the same primary, so even the intermediate peering (before pg-temp is fixed) is on the same osd.